compiler/rustc_expand/src/expand.rs

```diff
@@ -1260,25 +1260,33 @@ impl InvocationCollectorNode for P<ast::Item> {
         res
     }
     fn declared_names(&self) -> Vec<Ident> {
-        if let ItemKind::Use(ut) = &self.kind {
-            fn collect_use_tree_leaves(ut: &ast::UseTree, idents: &mut Vec<Ident>) {
-                match &ut.kind {
-                    ast::UseTreeKind::Glob => {}
-                    ast::UseTreeKind::Simple(_) => idents.push(ut.ident()),
-                    ast::UseTreeKind::Nested { items, .. } => {
-                        for (ut, _) in items {
-                            collect_use_tree_leaves(ut, idents);
+        struct ItemNameVisitor(Vec<Ident>);
```

Comment:
What happens in the case of a double or more stripped items? Won't this visitor and method record them twice?

```rust
#![allow(unexpected_cfgs)]

#[cfg(foo)]
mod fs {
    #[cfg(bar)]
    mod unix {
        fn fchown() {}
    )
}


fn main() {
    fs::unix::chown();
}
```

Reply:
This gives you

```
error[E0433]: failed to resolve: use of undeclared crate or module `fs`
  --> f72.rs:13:5
   |
13 |     fs::unix::chown();
   |     ^^ use of undeclared crate or module `fs`
   |
note: found an item that was configured out
  --> f72.rs:4:5
   |
4  | mod fs {
   |     ^^
note: the item is gated here
  --> f72.rs:3:1
   |
3  | #[cfg(foo)]
   | ^^^^^^^^^^^
```

because `fs` is not found so that's the only one that would be checked for (not `unix` or `chown`).

If you check for `unix`, it *will* point at `cfg(foo)`, which is not enough to get `unix` back, but gets you on the way to it:

```
error[E0433]: failed to resolve: use of undeclared crate or module `unix`
  --> f72.rs:14:5
   |
14 |     unix::chown();
   |     ^^^^ use of undeclared crate or module `unix`
   |
note: found an item that was configured out
  --> f72.rs:6:9
   |
6  |     mod unix {
   |         ^^^^
note: the item is gated here
  --> f72.rs:3:1
   |
3  | #[cfg(foo)]
   | ^^^^^^^^^^^
```

In testing, I noticed that we don't point at `fs::unix` when trying `unix` directly and it is behind a `cfg` (the second error):

```
error[E0433]: failed to resolve: could not find `unix` in `fs`
  --> f72.rs:13:9
   |
13 |     fs::unix::chown();
   |         ^^^^ could not find `unix` in `fs`
   |
note: found an item that was configured out
  --> f72.rs:6:9
   |
6  |     mod unix {
   |         ^^^^
note: the item is gated here
  --> f72.rs:5:5
   |
5  |     #[cfg(bar)]
   |     ^^^^^^^^^^^

error[E0433]: failed to resolve: use of undeclared crate or module `unix`
  --> f72.rs:14:5
   |
14 |     unix::chown();
   |     ^^^^ use of undeclared crate or module `unix`
```

If it isn't `cfg`d out *and* it's public, we do provide this suggestion:

```
error[E0433]: failed to resolve: use of undeclared crate or module `unix`
  --> f72.rs:14:5
   |
14 |     unix::chown();
   |     ^^^^ use of undeclared crate or module `unix`
   |
help: consider importing this module
   |
4  + use fs::unix;
   |
```

but if it is private and crate local maybe we should at least point at it (we don't):

```
error[E0433]: failed to resolve: use of undeclared crate or module `unix`
  --> f72.rs:14:5
   |
14 |     unix::chown();
   |     ^^^^ use of undeclared crate or module `unix`
```

Reply:
Changed to behave as you expected.